pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +266,178 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
 
 	return allErrs
 }
+
+// ValidateCSINodeInfo validates a CSINodeInfo.
+func ValidateCSINodeInfo(csiNodeInfo *storage.CSINodeInfo) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&csiNodeInfo.ObjectMeta, false, apivalidation.ValidateClassName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateCSINodeInfoSpec(&csiNodeInfo.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, validateCSINodeInfoStatus(&csiNodeInfo.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidateCSINodeInfoUpdate validates a CSINodeInfo.
+func ValidateCSINodeInfoUpdate(new, old *storage.CSINodeInfo) field.ErrorList {
+	allErrs := ValidateCSINodeInfo(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
```

Comment:
im not sure that the spec is completely immutable. Especially since we have plans to bootstrap `CSINodeInfo` on node startup then fill in the details later in a reconcile loop

Reply:
Spec gets updated anytime a driver is added/removed

Reply:
spec does not change on unregistration of Driver. The update is in `Available=false` (which I believe we've changed slightly). Not sure what you're referring to on "delete," I assume if you specifically delete the spec it will be gone.

Reply:
yes, we will unregister the driver and remove it from the object when the driver socket is removed (and it is picked up by the plugin registration mechanism).

For your second question I'm not aware of any cases where we will want to update `NodeID` or `TopologyKeys` after the spec is already created but I'm not as involved in those

Reply:
I think we can require that the driver unregister and re-register if they need to change nodeid or topologykeys

Reply:
Done